pkg/util/rowcodec/encoder.go

```diff
@@ -213,3 +221,36 @@ func encodeValueDatum(loc *time.Location, d *types.Datum, buffer []byte) (nBuffe
 	nBuffer = buffer
 	return
 }
+
+// Checksum is used to calculate and append checksum data into the raw bytes
+type Checksum interface {
+	encode(encoder *Encoder, buf []byte) ([]byte, error)
+}
+
+// NoChecksum indicate no checksum encoded into the raw bytes
+type NoChecksum struct{}
+
+func (NoChecksum) encode(encoder *Encoder, buf []byte) ([]byte, error) {
+	encoder.flags &^= rowFlagChecksum // revert checksum flag
+	return encoder.toBytes(buf), nil
+}
+
+const checksumVersionRaw byte = 1
+
+// RawChecksum indicates encode the raw bytes checksum and append it to the raw bytes.
+type RawChecksum struct {
+	Key []byte
+}
+
+func (c RawChecksum) encode(encoder *Encoder, buf []byte) ([]byte, error) {
+	encoder.flags |= rowFlagChecksum
+	encoder.checksumHeader &^= checksumFlagExtra   // revert extra checksum flag
+	encoder.checksumHeader &^= checksumMaskVersion // revert checksum version
+	encoder.checksumHeader |= checksumVersionRaw   // set checksum version
+	valueBytes := encoder.toBytes(buf)
+	valueBytes = append(valueBytes, encoder.checksumHeader)
+	encoder.checksum1 = crc32.Checksum(valueBytes, crc32.IEEETable)
+	encoder.checksum1 = crc32.Update(encoder.checksum1, crc32.IEEETable, c.Key)
```

Comment:
Does the calculation of `checksum2` need to be changed?

Reply:
There is no need to calculate the `checksum2`, since the checksum is calculated based on the raw bytes, no other factors.

The field `checksum2` is still keeped, only used to read out the previous version write extra checksum, for upgrade compatibility. Actually, this field won't be used since `tidb_row_checksum()` calculate the column-level checksum in the realtime by using the latest schema.